### PR TITLE
add a recipe step to create scope for openid

### PR DESCRIPTION
**Pull request by MichaelPetrinolis**
eg.
create a scope to validate external tokens on an IdP
```json
    {
      "name": "OpenIdScope",
      "Description": "A scope to provide audience for remote clients",
      "DisplayName": "External Audience Scope",
      "ScopeName": "[js: variables('scopeName')]",
      "Resources": "[js: variables('externalAudience')]"
    }
```

setup JWT bearer token to validate tokens from an external IdP
```json
    {
      "name": "OpenIdValidationSettings",
      "Audience": "[js: variables('externalAudience')]",
      "Authority": "https://idp.domain.com"
    }
```

### CODE REVIEW

**Comment by agriffard** [comment]
Can you please include your examples in documentation?

**Comment by MichaelPetrinolis** [comment]
@PinpointTownes  you can define variables in a recipe and use them later. You can also provide them from code if you run the recipe manually. Useful for SaaS implementations :)

**Comment by jtkech** [comment]
@PinpointTownes, as @MichaelPetrinolis said, LGTM.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdScopeStep.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+        public async Task ExecuteAsync(RecipeExecutionContext context)
+        {
+            if (!string.Equals(context.Name, "OpenIdScope", StringComparison.OrdinalIgnoreCase))
+            {
+                return;
+            }
+
+            var model = context.Step.ToObject<OpenIdScopeStepViewModel>();
+
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdScopeStep.cs]: Remove extra line

File: src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdValidationSettingsStep.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+        public async Task ExecuteAsync(RecipeExecutionContext context)
+        {
+            if (!string.Equals(context.Name, nameof(OpenIdValidationSettings), StringComparison.OrdinalIgnoreCase))
+            {
+                return;
+            }
+
+            var model = context.Step.ToObject<OpenIdValidationSettingsStepModel>();
+
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdValidationSettingsStep.cs]: Remove extra line

File: src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdValidationSettingsStep.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            var settings = await _validationService.GetSettingsAsync();
+            settings.Tenant = model.Tenant;
+            settings.Authority = !string.IsNullOrEmpty(model.Authority) ? new Uri(model.Authority, UriKind.Absolute) : null;
+            settings.Audience = model.Audience;
+
+            await _validationService.UpdateSettingsAsync(settings);
+        }
+    }
+
+    public class OpenIdValidationSettingsStepModel
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdValidationSettingsStep.cs]: Move it to a new file

File: src/docs/reference/modules/OpenId/README.md [hunk 4]
```diff
@@ -131,6 +131,30 @@ A sample of OpenID Connect App recipe step:
 }
 ```
 
+### OpenID Connect Scopes Configuration
+
+OpenID Connect scopes can be set through OpenID Connect Scopes menu in the admin dashboard (through the Management Interface feature) and also through a recipe step.
+
+OpenID Connect scopes require the following configuration.
```
Comment (hishamco) [inline src/docs/reference/modules/OpenId/README.md]: Scopes

File: src/docs/reference/modules/OpenId/README.md [hunk 5]
```diff
@@ -131,6 +131,30 @@ A sample of OpenID Connect App recipe step:
...
+
+- Name: Unique name of the scope.
```
Comment (hishamco) [inline src/docs/reference/modules/OpenId/README.md]: Using table will be better

File: src/docs/reference/modules/OpenId/README.md [hunk 6]
```diff
@@ -214,6 +238,22 @@ To use the certificate on an Azure hosted site.
   - Configure the validation feature to transparently use the server configuration of another tenant, which has the authorization server feature enabled.
 - Validates token by a remote server supporting JWT and OpenID Connect discovery.
 
+Token Validation require the following configuration.
+
+- Authorization server tenant: The tenant that runs OpenID Connect Server. If none is selected, then the following properties must be provided.
```
Comment (hishamco) [inline src/docs/reference/modules/OpenId/README.md]: Use table

File: src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdValidationSettingsStepModel.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+namespace OrchardCore.OpenId.Recipes
+{
+    public class OpenIdValidationSettingsStepModel
+    {
+
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdValidationSettingsStepModel.cs]: Remove this line

File: src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdServerSettingsStepModel.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
+using System.ComponentModel.DataAnnotations;
+using System.Security.Cryptography.X509Certificates;
+using static OrchardCore.OpenId.Settings.OpenIdServerSettings;
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdServerSettingsStepModel.cs]: Why we need this?

File: src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdValidationSettingsStep.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+using System.ComponentModel.DataAnnotations;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using OrchardCore.OpenId.Services;
+using OrchardCore.OpenId.Settings;
+using OrchardCore.Recipes.Models;
+using OrchardCore.Recipes.Services;
+using static OpenIddict.Abstractions.OpenIddictConstants;
+using static OrchardCore.OpenId.Settings.OpenIdServerSettings;
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdValidationSettingsStep.cs]: Why we need these?

File: src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdServerSettingsStepModel.cs (hunk 8 above)
Comment (MichaelPetrinolis) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdServerSettingsStepModel.cs]: The TokenFormat enum is defined in OpenIdServerSettings class, so the using static to the class is needed.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdValidationSettingsStep.cs (hunk 9 above)
Comment (MichaelPetrinolis) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdValidationSettingsStep.cs]: we don't

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdValidationSettingsStep.cs]: Just remove theme

File: src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdServerSettingsStepModel.cs (hunk 8 above)
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdServerSettingsStepModel.cs]: Ok, coz I didn't see `OpenIdServerSettings` usage in the code

File: src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdValidationSettingsStep.cs (hunk 9 above)
Comment (MichaelPetrinolis) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdValidationSettingsStep.cs]: i have already done so ;)

**Comment by sebastienros** [comment]
I went ahead and resolved the conflicts. I assume if it builds it should be fine.

**Comment by sebastienros** [comment]
@MichaelPetrinolis but if you could check that I did it correctly I would appreciate, thanks

File: src/OrchardCore.Modules/OrchardCore.OpenId/Startup.cs [hunk 10]
```diff
@@ -170,11 +170,65 @@ public override void ConfigureServices(IServiceCollection services)
         }
     }
 
+    [Feature(OpenIdConstants.Features.Management)]
```
Comment (kevinchalet) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Startup.cs]: @sebastienros bad merge?

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Startup.cs]: What I am asking, probably

Comment (kevinchalet) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Startup.cs]: None of the proposed changes makes sense, so I'd revert the entire file changes.

Comment (MichaelPetrinolis) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Startup.cs]: ok, reverted changes in startup.cs and added the recipe step handlers

**Pull request merged**